app/src/main/java/org/wikipedia/descriptions/DescriptionEditFragment.kt

```diff
@@ -0,0 +1,436 @@
+                }
+                ServiceFactory.get(wikiData).postDescriptionEdit(languageCode, languageCode, pageTitle.wikiSite.dbName(),
+                        pageTitle.prefixedText, binding.fragmentDescriptionEditView.description!!, comment, editToken,
+                        if (AccountUtil.isLoggedIn) "user" else null)
+            }
+        }
+
+        private fun editFailed(caught: Throwable, logError: Boolean) {
+            binding.fragmentDescriptionEditView.setSaveState(false)
+            FeedbackUtil.showError(requireActivity(), caught)
+            L.e(caught)
+            if (logError) {
+                funnel.logError(caught.message)
+            }
+            SuggestedEditsFunnel.get().failure(action)
+        }
+
+        override fun onHelpClick() {
+            FeedbackUtil.showAndroidAppEditingFAQ(requireContext())
+        }
+
+        override fun onCancelClick() {
+            if (binding.fragmentDescriptionEditView.showingReviewContent()) {
+                binding.fragmentDescriptionEditView.loadReviewContent(false)
+            } else {
+                DeviceUtil.hideSoftKeyboard(requireActivity())
+                requireActivity().onBackPressed()
+            }
+        }
+
+        override fun onBottomBarClick() {
+            callback()?.onBottomBarContainerClicked(action)
+        }
+
+        override fun onVoiceInputClick() {
+            val intent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH)
+                    .putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM)
+            try {
+                startActivityForResult(intent, Constants.ACTIVITY_REQUEST_VOICE_SEARCH)
+            } catch (a: ActivityNotFoundException) {
+                FeedbackUtil.showMessage(requireActivity(), R.string.error_voice_search_not_available)
+            }
+        }
+    }
+
+    private fun updateDescriptionInArticle(articleText: String, newDescription: String): String {
+        return if (templateParsePattern.matcher(articleText).find()) {
+            // update existing description template
+            articleText.replaceFirst(TEMPLATE_PARSE_REGEX.toRegex(), "$1$newDescription$3")
+        } else {
+            // add new description template
+            """{{${DESCRIPTION_TEMPLATES[0]}|$newDescription}}$articleText""".trimIndent()
+        }
+    }
+
+    companion object {
+        private val DESCRIPTION_TEMPLATES = arrayOf("Short description", "SHORTDESC")
+        private const val ARG_TITLE = "title"
+        private const val ARG_REVIEWING = "inReviewing"
+        private const val ARG_DESCRIPTION = "description"
+        private const val ARG_HIGHLIGHT_TEXT = "highlightText"
+        private const val ARG_ACTION = "action"
+        private const val ARG_SOURCE_SUMMARY = "sourceSummary"
+        private const val ARG_TARGET_SUMMARY = "targetSummary"
+        const val TEMPLATE_PARSE_REGEX = "(\\{\\{[Ss]hort description\\|(?:1=)?)([^}|]+)([^}]*}})"
```

Comment:
Oops, looks like this change causes the crash when opening the `DescriptionEditFragment`, maybe we should add the slashes back and maybe add `Suppress()` for that.